bugtool/cmd/root.go

```diff
@@ -65,26 +66,42 @@ var (
 )
 
 func init() {
+	BugtoolRootCmd.Flags().BoolVar(&archive, "archive", true, "Create archive")
 	BugtoolRootCmd.Flags().BoolVar(&serve, "serve", false, "Start HTTP server to serve static files")
+	BugtoolRootCmd.Flags().BoolVar(&k8s, "k8s-mode", false, "Require Kubernetes pods to be found or fail")
 	BugtoolRootCmd.Flags().IntVarP(&port, "port", "p", 4444, "Port to use for the HTTP server, (default 4444)")
 	BugtoolRootCmd.Flags().StringVarP(&dumpPath, "tmp", "t", "/tmp", "Path to store extracted files")
 	BugtoolRootCmd.Flags().StringVarP(&host, "host", "H", "", "URI to server-side API")
 	BugtoolRootCmd.Flags().StringVarP(&k8sNamespace, "k8s-namespace", "", "kube-system", "Kubernetes namespace for Cilium pod")
 	BugtoolRootCmd.Flags().StringVarP(&k8sLabel, "k8s-label", "", "k8s-app=cilium", "Kubernetes label for Cilium pod")
 }
 
-func runTool() {
-	// Search for a Cilium pod and if one is found prefix all of the
-	// commands to use the pod(s).
+func setup() {
+	// By default try to pick either Kubernetes or non-k8s (host mode). If
+	// we find Cilium pod(s) then it's k8s-mode otherwise host mode.
+	// Passing extra flags can override the default.
 	var err error
 	k8sPods, err = getCiliumPods(k8sNamespace, k8sLabel)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Unable to retrieve cilium logs from kubernetes, skipping... %s\n", err)
+	switch {
+	case k8s:
+		// When the k8s flag is set, perform extra checks that we actually do have pods or fail.
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\nFailed to find pods, is kube-apiserver running?\n", err)
+			os.Exit(1)
+		}
+		if len(k8sPods) < 1 {
+			fmt.Fprintf(os.Stderr, "Found no pods, is kube-apiserver running?\n")
+			os.Exit(1)
+		}
+	case os.Getuid() != 0 && len(k8sPods) == 0:
```

Comment:
Why is `len(k8sPods) == 0` relevant here?

Reply:
The warning should only show up if we are not root and there are no pods. If there are any pods that will affect how the commands are executed. Which mean that it's likely inside a container with root access since Cilium deployments require `privileged=true`. The warning would not make any sense if we found pods.